File: src/reimpl.c
// Copyright (c) 2026 渟雲. All rights reserved.
#include "./common.h"

__int64 _kascii_stricmp(const char* a1, const char* a2) {
  int v4;  // r8d
  int v5;  // edx
  int v6;  // r9d
  int v7;  // eax

  do {
    v4 = *a1++;
    v5 = *a2;
    v6 = v4 + 32;
    if ((unsigned int)(v4 - 65) > 0x19) v6 = v4;
    v7 = v5 + 32;
    ++a2;
    if ((unsigned int)(v5 - 65) > 0x19) v7 = v5;
  } while (v6 && v6 == v7);
  return (unsigned int)(v6 - v7);
}

int kstricmp(const char* Str1, const char* Str2) {
  return (int)_kascii_stricmp(Str1, Str2);
}

int kwcsicmp(const wchar_t* Str1, const wchar_t* Str2) {
  const wchar_t* v2;    // r10
  signed __int64 v3;    // r9
  unsigned __int16 v4;  // r8
  unsigned __int16 v5;  // cx
  unsigned __int16 v6;  // dx
  unsigned __int16 v7;  // r8

  v2 = Str2;
  v3 = (char*)Str1 - (char*)Str2;
  do {
    v4 = *(const wchar_t*)((char*)v2 + v3);
    v5 = *v2++;
    v6 = v4 + 32;
    if ((unsigned __int16)(v4 - 65) > 0x19u) v6 = v4;
    v7 = v5 + 32;
    if ((unsigned __int16)(v5 - 65) > 0x19u) v7 = v5;
  } while (v6 && v6 == v7);
  return v6 - v7;
}

size_t __cdecl kwcslen(const wchar_t* Str) {
  __m128i* i;           // rdx
  unsigned __int64 v2;  // r9
  const wchar_t* v3;    // rax
  size_t v4;            // rdx

  i = (__m128i*)Str;
  if (((uintptr_t)Str & 1) != 0) {
    while (i->m128i_i16[0]) i = (__m128i*)((char*)i + 2);
    return ((char*)i - (char*)Str) >> 1;
  }
  v2 = ((16LL - ((uintptr_t)Str & 0xF)) &
        (unsigned __int64)-(__int64)(((uintptr_t)Str & 0xF) != 0)) >>
       1;
  v3 = &Str[v2];
  if (Str != v3) {
    do {
      if (!i->m128i_i16[0]) break;
      i = (__m128i*)((char*)i + 2);
    } while (i != (__m128i*)v3);
  }
  v4 = ((char*)i - (char*)Str) >> 1;
  if (v4 == v2) {
    for (i = (__m128i*)&Str[v4];
         !_mm_movemask_epi8(_mm_cmpeq_epi16(_mm_setzero_si128(), *i)); ++i);
    while (i->m128i_i16[0]) i = (__m128i*)((char*)i + 2);
    return ((char*)i - (char*)Str) >> 1;
  }
  return v4;
}

ULONG g_PebOffset = 0;

ULONG GetPebOffset() {
  if (g_PebOffset != 0) return g_PebOffset;

  UNICODE_STRING routineName = RTL_CONSTANT_STRING(L"PsGetProcessPeb");

  PUCHAR pFunc = (PUCHAR)MmGetSystemRoutineAddress(&routineName);
  if (!pFunc) {
    return 0;
  }

  // 48 8B 81 ?? ?? ?? ?? mov     rax, [rcx+2E0h]
  for (int i = 0; i < 0x10; i++) {
    if (pFunc[i] == 0x48 && pFunc[i + 1] == 0x8B && pFunc[i + 2] == 0x81) {
      ULONG offset = *(PULONG)(pFunc + i + 3);
      // just guessing the range
      if (offset > 0x10 && offset <= 0x1000) {
        g_PebOffset = offset;
        return offset;
      }
    }
  }

  return 0;
}

PPEB PsGetProcessPebTrick(PEPROCESS Process) {
  ULONG offset = GetPebOffset();
  return *(PPEB*)((PUCHAR)Process + offset);
}

ULONG g_ImageFileNameOffset = 0;

ULONG GetImageFileNameOffset() {
  if (g_ImageFileNameOffset != 0) return g_ImageFileNameOffset;

  UNICODE_STRING routineName = RTL_CONSTANT_STRING(L"PsGetProcessImageFileName");

  PUCHAR pFunc = (PUCHAR)MmGetSystemRoutineAddress(&routineName);
  if (!pFunc) {
    return 0;
  }

  // 48 8D 81 ?? ?? ?? ?? lea     rax, [rcx+338h]
  for (int i = 0; i < 0x10; i++) {
    if (pFunc[i] == 0x48 && pFunc[i + 1] == 0x8D && pFunc[i + 2] == 0x81) {
      ULONG offset = *(PULONG)(pFunc + i + 3);
      // just guessing the range
      if (offset > 0x10 && offset <= 0x1000) {
        g_ImageFileNameOffset = offset;
        return offset;
      }
    }
  }

  return 0;
}

PCHAR PsGetProcessImageFileNameTrick(PEPROCESS Process) {
  ULONG offset = GetImageFileNameOffset();
  return (PCHAR)((PUCHAR)Process + offset);
}

ULONG g_ProcessIdOffset = 0;

ULONG GetProcessIdOffset() {
  if (g_ProcessIdOffset != 0) return g_ProcessIdOffset;

  UNICODE_STRING routineName =
      RTL_CONSTANT_STRING(L"PsGetProcessId");

  PUCHAR pFunc = (PUCHAR)MmGetSystemRoutineAddress(&routineName);
  if (!pFunc) {
    return 0;
  }

  // 48 8B 81 ?? ?? ?? ?? mov     rax, [rcx+1D0h]
  for (int i = 0; i < 0x10; i++) {
    if (pFunc[i] == 0x48 && pFunc[i + 1] == 0x8B && pFunc[i + 2] == 0x81) {
      ULONG offset = *(PULONG)(pFunc + i + 3);
      // just guessing the range
      if (offset > 0x10 && offset <= 0x1000) {
        g_ProcessIdOffset = offset;
        return offset;
      }
    }
  }

  return 0;
}

HANDLE PsGetProcessIdTrick(PEPROCESS Process) {
  ULONG offset = GetProcessIdOffset();
  return *(HANDLE*)((PUCHAR)Process + offset);
}

ULONG g_ProcessExitStatusOffset = 0;

ULONG GetProcessExitStatusOffset() {
  if (g_ProcessExitStatusOffset != 0) return g_ProcessExitStatusOffset;

  UNICODE_STRING routineName = RTL_CONSTANT_STRING(L"PsGetProcessExitStatus");

  PUCHAR pFunc = (PUCHAR)MmGetSystemRoutineAddress(&routineName);
  if (!pFunc) {
    return 0;
  }

  // 8B 81 ?? ?? ?? ?? mov     eax, [rcx+554h]
  for (int i = 0; i < 0x10; i++) {
    if (pFunc[i] == 0x8B && pFunc[i + 1] == 0x81) {
      ULONG offset = *(PULONG)(pFunc + i + 2);
      // just guessing the range
      if (offset > 0x10 && offset <= 0x1000) {
        g_ProcessExitStatusOffset = offset;
        return offset;
      }
    }
  }

  return 0;
}

NTSTATUS PsGetProcessExitStatusTrick(PEPROCESS Process) {
  ULONG offset = GetProcessExitStatusOffset();
  return *(NTSTATUS*)((PUCHAR)Process + offset);
}

File: src/reimpl.rs
// Copyright (c) 2026 渟雲. All rights reserved.

use core::ffi::c_void;
use core::ptr::null_mut;
use iced_x86::{Code, Decoder, DecoderOptions, Mnemonic, OpKind, Register};

use crate::imports::resolve_kernel_export;

fn extract_load_disp32(func: *mut c_void, cap: usize) -> u32 {
    unsafe {
        if func.is_null() {
            return 0;
        }
        let data = core::slice::from_raw_parts(func as *const u8, 0x60);
        let mut dec = Decoder::with_ip(64, data, func as u64, DecoderOptions::NONE);
        for _ in 0..cap {
            let insn = dec.decode();
            if insn.code() == Code::INVALID {
                return 0;
            }
            if insn.mnemonic() != Mnemonic::Mov && insn.mnemonic() != Mnemonic::Lea {
                continue;
            }
            if insn.memory_base() == Register::None || insn.is_ip_rel_memory_operand() {
                continue;
            }
            if insn.op0_kind() != OpKind::Register {
                continue;
            }
            let base = insn.memory_base();
            if base == Register::RSP || base == Register::RBP {
                continue;
            }
            if insn.memory_index() != Register::None {
                continue;
            }
            let d = insn.memory_displacement64();
            if d > 0x10 && d <= 0x1000 {
                return d as u32;
            }
        }
        0
    }
}

static mut G_PEB_OFFSET: u32 = 0;

fn get_peb_offset() -> u32 {
    unsafe {
        if G_PEB_OFFSET != 0 {
            return G_PEB_OFFSET;
        }
        let off = extract_load_disp32(resolve_kernel_export(crate::hash!(b"PsGetProcessPeb")), 10);
        if off != 0 {
            G_PEB_OFFSET = off;
        }
        off
    }
}

pub fn ps_get_process_peb_trick(process: *mut c_void) -> *mut c_void {
    let off = get_peb_offset();
    if off == 0 {
        return null_mut();
    }
    crate::helpers::read_u64(process as *const u8, off as usize) as *mut c_void
}

static mut G_IMAGE_FILE_NAME_OFFSET: u32 = 0;

fn get_image_file_name_offset() -> u32 {
    unsafe {
        if G_IMAGE_FILE_NAME_OFFSET != 0 {
            return G_IMAGE_FILE_NAME_OFFSET;
        }
        let off = extract_load_disp32(
            resolve_kernel_export(crate::hash!(b"PsGetProcessImageFileName")),
            10,
        );
        if off != 0 {
            G_IMAGE_FILE_NAME_OFFSET = off;
        }
        off
    }
}

pub fn ps_get_process_image_file_name_trick(process: *mut c_void) -> *mut i8 {
    unsafe {
        let off = get_image_file_name_offset();
        if off == 0 {
            return null_mut();
        }
        (process as *const u8).add(off as usize) as *mut i8
    }
}

static mut G_PROCESS_ID_OFFSET: u32 = 0;

fn get_process_id_offset() -> u32 {
    unsafe {
        if G_PROCESS_ID_OFFSET != 0 {
            return G_PROCESS_ID_OFFSET;
        }
        let off = extract_load_disp32(resolve_kernel_export(crate::hash!(b"PsGetProcessId")), 10);
        if off != 0 {
            G_PROCESS_ID_OFFSET = off;
        }
        off
    }
}

pub fn ps_get_process_id_trick(process: *mut c_void) -> usize {
    let off = get_process_id_offset();
    if off == 0 {
        return 0;
    }
    crate::helpers::read_u64(process as *const u8, off as usize) as usize
}

static mut G_PROCESS_EXIT_STATUS_OFFSET: u32 = 0;

fn get_process_exit_status_offset() -> u32 {
    unsafe {
        if G_PROCESS_EXIT_STATUS_OFFSET != 0 {
            return G_PROCESS_EXIT_STATUS_OFFSET;
        }
        let off = extract_load_disp32(
            resolve_kernel_export(crate::hash!(b"PsGetProcessExitStatus")),
            10,
        );
        if off != 0 {
            G_PROCESS_EXIT_STATUS_OFFSET = off;
        }
        off
    }
}

pub fn ps_get_process_exit_status_trick(process: *mut c_void) -> i32 {
    let off = get_process_exit_status_offset();
    if off == 0 {
        return 0;
    }
    crate::helpers::read_u32(process as *const u8, off as usize) as i32
}

static mut G_PFN_BASE: u64 = 0;
static mut G_PTE_BASE: u64 = 0;

fn parse_bases() {
    unsafe {
        if G_PFN_BASE != 0 && G_PTE_BASE != 0 {
            return;
        }
        let func = resolve_kernel_export(crate::hash!(b"MmGetVirtualForPhysical"));
        if func.is_null() {
            return;
        }
        let data = core::slice::from_raw_parts(func as *const u8, 0x60);
        let mut dec = Decoder::with_ip(64, data, func as u64, DecoderOptions::NONE);
        let mut caps = [0u64; 6];
        let mut nc = 0usize;
        for _ in 0..16 {
            let insn = dec.decode();
            if insn.code() == Code::INVALID {
                break;
            }
            let c = match insn.code() {
                Code::Mov_r64_imm64 => Some(insn.immediate64()),
                Code::Mov_rm64_imm32 => Some(insn.immediate32() as i32 as i64 as u64),
                _ => {
                    if insn.mnemonic() == Mnemonic::Mov || insn.mnemonic() == Mnemonic::Lea {
                        if insn.is_ip_rel_memory_operand() && insn.op0_kind() == OpKind::Register {
                            let disp = insn.memory_displacement64() as i64;
                            let mut v = (insn.next_ip() as u64).wrapping_add_signed(disp);
                            if insn.mnemonic() == Mnemonic::Mov {
                                v = crate::helpers::read_u64(v as *const u8, 0);
                            }
                            if (v >> 47) as u16 == 0xFFFF {
                                Some(v)
                            } else {
                                None
                            }
                        } else {
                            None
                        }
                    } else {
                        None
                    }
                }
            };
            if let Some(c) = c {
                if c != 0 && nc < caps.len() {
                    caps[nc] = c;
                    nc += 1;
                }
            }
        }
        if nc == 0 {
            return;
        }
        let mut pte = 0u64;
        let mut pfn = 0u64;
        for i in 0..nc {
            if caps[i] & 0x0fff_ffff == 0 {
                pte = caps[i];
                if i > 0 {
                    pfn = caps[0];
                }
                break;
            }
        }
        if pte == 0 {
            pte = caps[0];
            pfn = if nc > 1 { caps[1] } else { 0 };
        } else if pfn == 0 {
            pfn = caps.iter().copied().find(|&c| c != pte).unwrap_or(0);
        }
        G_PTE_BASE = pte;
        G_PFN_BASE = pfn;
    }
}

pub fn mm_get_virtual_for_physical_trick(physical_address: u64) -> *mut c_void {
    unsafe {
        parse_bases();
        let pfn = physical_address >> 12;
        let offset = physical_address & 0xFFF;
        let val = crate::helpers::read_u64((G_PFN_BASE + pfn * 48) as *const u8, 0);
        let shifted = val << 25;
        let base2_shifted = G_PTE_BASE << 25;
        let diff = (shifted.wrapping_sub(base2_shifted) as i64) >> 16;
        (diff as u64 + offset) as *mut c_void
    }
}
